### PR TITLE
Backport newer uint256 types from Bitcoin

### PR DESCRIPTION
**Pull request by cyrossignol**
This updates the `uint256` type to Bitcoin's latest version which splits it into two types: `uint256` is just a basic blob type, and `arith_uint256` implements more specialized functionality for numeric operations and comparisons. I refactored the application to use the appropriate type for each instance.

This fixes serialization of `uint256` objects for big-endian platforms (#1196 ). For this, an upgrade is overkill, but the newer versions improve the safety of the code a bit and will make it easier to upgrade other areas of the application in the future.

Because the newer uint256.h code does not provide a `uint128` type anymore, I switched the type of `CBlockIndex::cpid` to `NN::Cpid`--it was only used here. This PR finishes backporting prerequisite strencodings utilities that @jamescowens started for #1526.

### CODE REVIEW

**Review by jamescowens — APPROVED** [review]

**Comment by jamescowens** [comment]
Looks good to me...

File: src/Makefile.am [hunk 1]
```diff
@@ -131,6 +132,7 @@ GRIDCOIN_CORE_H = \
 	ui_interface.h \
 	uint256.h \
 	util/memory.h \
+	util/strencodings.h
```
Comment (jamescowens) [inline src/Makefile.am]: missed a backslash. that is why Travis is barking

**Review by jamescowens — CHANGES_REQUESTED** [review]

File: src/Makefile.am (hunk 1 above)
Comment (cyrossignol) [reply src/Makefile.am]: Derp...fixed, thanks!

**Comment by cyrossignol** [comment]
Rebased and fixed a comparison issue for the stake modifier (the newer, byte-array-backed `uint256` from Bitcoin implements `operator<()` differently so it broke the stake modifier selection).

**Review by jamescowens — APPROVED** [review]

**Comment by cyrossignol** [comment]
Rebased to update `uint256` code added in #1559.

**Pull request merged**